Bottom: 775ae69335e53734a2445d133f498759cc3c047c
Top:    57c09c6d2735083838ace3d9ca92e7e96220fbb7
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-02-21 13:36:56 +0800

sample/xdp: make xdp tunnel sample works unconditonally

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/samples/bpf/xdp_tx_iptunnel_kern.c b/samples/bpf/xdp_tx_iptunnel_kern.c
index 0f4f6e8..89b24cb 100644
--- a/samples/bpf/xdp_tx_iptunnel_kern.c
+++ b/samples/bpf/xdp_tx_iptunnel_kern.c
@@ -78,7 +78,7 @@ static __always_inline int handle_ipv4(struct xdp_md *xdp)
 {
 	void *data_end = (void *)(long)xdp->data_end;
 	void *data = (void *)(long)xdp->data;
-	struct iptnl_info *tnl;
+	struct iptnl_info tnl = {0};
 	struct ethhdr *new_eth;
 	struct ethhdr *old_eth;
 	struct iphdr *iph = data + sizeof(struct ethhdr);
@@ -102,10 +102,16 @@ static __always_inline int handle_ipv4(struct xdp_md *xdp)
 	vip.dport = dport;
 	payload_len = ntohs(iph->tot_len);
 
+
+//	tnl.daddr.v4 = iph->daddr;
+//	tnl.saddr.v4 = iph->saddr;
+//	tnl.dmac = old_eth->h_dest;
+	#if 0
 	tnl = bpf_map_lookup_elem(&vip2tnl, &vip);
 	/* It only does v4-in-v4 */
 	if (!tnl || tnl->family != AF_INET)
 		return XDP_PASS;
+	#endif
 
 	/* The vip key is found.  Add an IP header and send it out */
 
@@ -124,7 +130,7 @@ static __always_inline int handle_ipv4(struct xdp_md *xdp)
 	    iph + 1 > data_end)
 		return XDP_DROP;
 
-	set_ethhdr(new_eth, old_eth, tnl, htons(ETH_P_IP));
+	set_ethhdr(new_eth, old_eth, &tnl, htons(ETH_P_IP));
 
 	iph->version = 4;
 	iph->ihl = sizeof(*iph) >> 2;
@@ -133,8 +139,8 @@ static __always_inline int handle_ipv4(struct xdp_md *xdp)
 	iph->check = 0;
 	iph->tos = 0;
 	iph->tot_len = htons(payload_len + sizeof(*iph));
-	iph->daddr = tnl->daddr.v4;
-	iph->saddr = tnl->saddr.v4;
+	iph->daddr = tnl.daddr.v4;
+	iph->saddr = tnl.saddr.v4;
 	iph->ttl = 8;
 
 	next_iph_u16 = (u16 *)iph;
